MudTable grouping

HClausing opened the pull request:
Grouping feature for MudTable, with Group Header and Footer Rows, allowing Summaries on both.

It works on recursive way, so it's possible to have unlimited inner groups.

Proposal for #1475 , and replacement for #1879.

Also added overflow-y:auto to mud-table-container css class to avoid rows getting outsite from table container when table has fixed height, or doesn't use fixed header neither horizontal scrollbar. Example here: https://try.mudblazor.com/snippet/GOcFYqlCavftHnBY

Reviews are always welcome.

HClausing committed:
Added recursive grouping behavior for MudTable<T>, with group header and footer templates.

Also added overflow-y:auto to mud-table-container css class to avoid rows getting outsite from table container when table has fixed height, or doesn't use fixed header neither horizontal scrollbar

HClausing committed:
Fix for build 20210618.1 test failure

henon commented:
Here is some feedback:

I think we should remove the checkbox from the footer

![image](https://user-images.githubusercontent.com/44090/122674389-39a0a280-d1d5-11eb-8f8a-48692f374b33.png)


Also I am not entirely satisfied how the header row and the footer row look. Right now a footer is immediately followed by a header without any spacing between them. This is quite puzzling on first glance.

I suggest we look how other toolkits do it to find a good default look for row groups with header and footer.

henon commented:
The presentation problem is even more apparent when you look at multi-level grouping:

![image](https://user-images.githubusercontent.com/44090/122674705-a6686c80-d1d6-11eb-9c63-d44bf7464651.png)

HClausing commented:
Thanks. I appreciate your feedback, and I totally agree.

I'll try find a more elegant way to present the rows when using group footer.

HClausing committed:
Removed <MudCheckBox @bind-Checked="IsChecked" /> from Group Footer even if MultiSelect is true.

henon commented:
Would you mind renaming MudTGroupRow to MudTableGroupRow ?

HClausing commented:
Good ideia. I'll rename
Thanks.

HClausing committed:
- Renamed MudTGroupRow to MudTableGroupRow;
 - Removed MultiSelection checkbox from Footer Row;
 - Removed gray background stylee on Footer Rows from Docs.

HClausing committed:
- Added Indentation parameter;
 - Added Expandable parameter.
 - Added tests for Indentation and Expandable.

HClausing commented:
The current MudTable Grouping behavior (every group header is gray but my gif is under low color quality, sorry):

![grp](https://user-images.githubusercontent.com/15158923/123313628-12611280-d500-11eb-859d-bce681c2f1b4.gif)

henon commented:
It looks a lot better already. I love the expandable function. However I think like this it would be even better (I quickly threw this sketch together):

![image](https://user-images.githubusercontent.com/44090/123316538-3dae2680-d52d-11eb-9fe9-368668c97c8a.png)

henon commented:
Actually it is not much work because we have the `Group` property on `Element`. I just pushed an update to the group names so that they are better suited for this example.

HClausing commented:
> It looks a lot better already. I love the expandable function. However I think like this it would be even better (I quickly threw this sketch together):
> 
> ![image](https://user-images.githubusercontent.com/44090/123316538-3dae2680-d52d-11eb-9fe9-368668c97c8a.png)
> 
> I am talking about the order of the expansion icon and the checkbox. Also the example would be so much more intuitive if we would group by the big element groups: Metals, Non-metals, etc. I know it is more work, but it would be a wow-effect.

Oh really, I've inverted the order 🙄. 

I'm working on it, and the biggest challenge is about to get the ensure the correct indentation without break the column width.

HClausing committed:
Fixed GroupExpander Toogle and MultiSelection Checkbox order;
Changed GroupExpander Toogle and MultiSelection Checkbox to the sasme Tr column;
Fixed some misalignment when Indent is active;
Fixed header row that was being rendered as MudTd instead of MudTr
Some updates on examples.

HClausing commented:
After sosme days testing and rendering, I've made some progress:

![image](https://user-images.githubusercontent.com/15158923/123686961-be6b6c00-d826-11eb-9577-be889d24e0c9.png)

Now, about the empty space between expander toggle / checkbox and first column, I guess a better looking will be only possible when the PR #1110 ne merged, because will be possible to override the MudTd of the first column adding these components before.

![image](https://user-images.githubusercontent.com/15158923/123687366-3a65b400-d827-11eb-9e06-5f60a0003b0a.png)

henon commented:
It looks awesome!

HClausing commented:
> It looks awesome!

Thanks!

henon commented:
@mikes-gh @Garderoben do you think this is OK to be merged now?

Garderoben commented:
> @mikes-gh @Garderoben do you think this is OK to be merged now?

Design wise, YES!

henon commented:
Many many thanks!

Yomodo commented:
Wonderful! Thanks.

henon commented on src/MudBlazor/Components/Table/MudTable.razor:
```diff
@@ -123,4 +137,39 @@
+        @<text>
+            @foreach (T item in source)
+            {
+                var rowClass = new CssBuilder(RowClass).AddClass(RowClassFunc?.Invoke(item, rowIndex)).AddClass(customClass, ! string.IsNullOrEmpty("mud-table-row-group-indented-1")).Build();
```
The bug is in this line. @HClausing can you remember what your intention was? the expression `! string.IsNullOrEmpty("mud-table-row-group-indented-1")` evaluates to `true` which will always add the `customClass`.

HClausing replied:
I'm trying to remember. 

Maybe it's related to indent level, but implementing by this way it doens't make any sense, or I've lost the original plan, or else I didn't finished this feature.

Looking at css files, I've found this:

```
tr.mud-table-row-group-indented-1 td:first-child {
    padding-left: 48px !important;
}

tr.mud-table-row-group-indented-2 td:first-child {
    padding-left: 96px !important;
}

tr.mud-table-row-group-indented-3 td:first-child {
    padding-left: 144px !important;
}

tr.mud-table-row-group-indented-4 td:first-child {
    padding-left: 192px !important;
}

tr.mud-table-row-group-indented-5 td:first-child {
    padding-left: 240px !important;
}

```

... and just the level 1 is in use. I guess the padding/indentation is being applying by another way.

I'm sorry for this.

henon replied:
Don't worry, mistakes happen.

henon commented:
@HClausing The guys from PVS found a bug in razor that was added in this PR, see https://pvs-studio.com/en/blog/posts/csharp/1023/

I have commented above accordingly.

HClausing commented on src/MudBlazor/Components/Table/MudTableGroupRow.razor:
```diff
@@ -0,0 +1,78 @@
+    }
+    else
+    {
+        @Table?.RenderRows(Items, (GroupDefinition?.Indentation ?? false) ? $"mud-table-row-group-indented-{GroupDefinition?.Level - 1}" : null, GroupDefinition?.IsThisOrParentExpandable ?? false)
```
@henon , please take a look here.

I guess that condition should be removed on .AddClass() (MudTable.razor, line 151).

henon replied:
ok, great. as you know this code best would you PR the fix and tag me? thanks.

HClausing replied:
Yes, for sure! Would be a problem if I take a look at this only on the next week ?

henon replied:
No problem, this is not urgent.

HClausing replied:
I'm working on it now, and figured that the correct condition is `!string.IsNullOrEmpty(customClass)` (argument).

HClausing replied:
I've opened the #6374 pul request. Thanks for let me know about this.
